src/main/java/com/bobocode/hoverla/bring/context/ApplicationContextImpl.java

```diff
@@ -1,60 +1,152 @@
 package com.bobocode.hoverla.bring.context;
 
+import com.bobocode.hoverla.bring.exception.NoSuchBeanException;
+import com.bobocode.hoverla.bring.exception.NoUniqueBeanException;
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Table;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.Objects;
+
 
 /**
- * Example of ApplicationContext implementation, can be removed or reimplemented when actual development starts
+ * Implementation of ApplicationContext
  */
+@Slf4j
 public class ApplicationContextImpl implements ApplicationContext {
+    private Table<String, Class<?>, BeanDefinition> context = HashBasedTable.create();
+    private final String BEAN_TYPE_MUST_BE_NOT_NULL_MESSAGE = "The argument [beanType] must be not null";
+    private final String NO_SUCH_BEAN_EXCEPTION_MESSAGE = "Bean with provided name/type [%s] not found in the context";
+
+    /**
+     *  Scanners scan application to define beans configs and map them into {@link BeanDefinition} lists.
+     *  Received list of bean definitions validates by validator to avoid duplicates and other problems (see {@link BeanValidator}).
+     *  Beans definitions list maps to container structure. It's Table<String, Class<?>, BeanDefinition>, where
+     *      - row key is a name of bean
+     *      - column key is a type of bean
+     *      - mapped value is bean definition
+     *  Then context passes to {@link BeanInitializer} for beans instances initialization
+     *
+     * @param scanners      list of scanners for
+     * @param validator     bean definition validator
+     * @param initializer   bean initializer
+     */
     public ApplicationContextImpl(List<BeanScanner> scanners,
                                   BeanValidator validator,
                                   BeanInitializer initializer) {
-        /* 1) Scanners get all beans
-           2) Merge List of bean definitions into one List
-           3) Send beans list to validator to check duplicates
-           4) Convert bean list to map
-           5) Send beans map to initializer to instantiate beans and inject dependencies
-         */
+
+        List<BeanDefinition> beanDefinitionList = scanningApplication(scanners);
+        validate(validator, beanDefinitionList);
+        contextCreation(beanDefinitionList);
+        beansInstantiation(initializer);
+
+        log.info("Application context initialized");
+    }
+
+    private List<BeanDefinition> scanningApplication(List<BeanScanner> scanners) {
+        List<BeanDefinition> beanDefinitionList = new ArrayList<>();
+        log.info("Scanning application for finding bean definitions");
```

Comment:
I would rephrase this log message to `Started bean definition scanning`

Reply:
@NazarSmith228 I don't like "started", because when I see message like that than I expect message "ended" appears. 
And one more thing - it's not started yet, it WILL be started.
What do you think about "Bean definition scanning" ?

Reply:
@nadya-gainutsa then let's log nothing here, since scanners will let us know that they started